File: src/keyboard.ts
/** @noSelfInFile */

export const moduleName = 'keyboard'
export const log = hs.logger.new(moduleName, 'info')

export function useSquirrel(): void {
  const currentID = hs.keycodes.currentSourceID()
  const currentMethod = hs.keycodes.currentMethod()
  const isSquirrel = (currentID && currentID.includes('Squirrel')) || (currentMethod && currentMethod.includes('Squirrel'))

  if (isSquirrel) {
    // If Squirrel is already active, send Control+Shift+Alt+F2.
    // This triggers the Rime configuration:
    // `- { when: always, accept: "Control+Shift+Alt+F2", unset_option: ascii_mode }`
    // to force switch from ASCII mode back to Chinese mode.
    hs.eventtap.keyStroke(['ctrl', 'shift', 'alt'], 'f2')
  }
  else {
    const methods = hs.keycodes.methods()
    const squirrel = methods.find(method => method.includes('Squirrel'))
    if (squirrel) {
      hs.keycodes.setMethod(squirrel)
    }
  }
}

export function bindSquirrelHotkey(): void {
  log.i('Binding squirrel hotkey...')
  hs.hotkey.bind(['cmd'], 'escape', useSquirrel)
}

function init(): void {
  log.i(`Initializing [${moduleName}] module...`)
  bindSquirrelHotkey()
}

export function apply(): void {
  init()
}

File: keyboard.lua
local ____lualib = require("lualib_bundle")
local __TS__StringIncludes = ____lualib.__TS__StringIncludes
local __TS__ArrayFind = ____lualib.__TS__ArrayFind
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["9"] = 3,["10"] = 4,["11"] = 6,["12"] = 7,["13"] = 8,["14"] = 8,["15"] = 8,["16"] = 8,["17"] = 9,["18"] = 10,["20"] = 6,["21"] = 14,["22"] = 15,["23"] = 16,["24"] = 14,["25"] = 19,["26"] = 20,["27"] = 21,["28"] = 19,["29"] = 24,["30"] = 25,["31"] = 24});
local ____exports = {}
---
-- @noSelfInFile
____exports.moduleName = "keyboard"
____exports.log = hs.logger.new(____exports.moduleName, "info")
function ____exports.useSquirrel()
    local methods = hs.keycodes.methods()
    local squirrel = __TS__ArrayFind(
        methods,
        function(____, method) return __TS__StringIncludes(method, "Squirrel") end
    )
    if squirrel then
        hs.keycodes.setMethod(squirrel)
    end
end
function ____exports.bindSquirrelHotkey()
    ____exports.log.i("Binding squirrel hotkey...")
    hs.hotkey.bind({"cmd"}, "escape", ____exports.useSquirrel)
end
local function init()
    ____exports.log.i(("Initializing [" .. ____exports.moduleName) .. "] module...")
    ____exports.bindSquirrelHotkey()
end
function ____exports.apply()
    init()
end
return ____exports
